From d0e3b47e86eb1c169fccc90b824a327de23a0bf4 Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@gmail.com>
Date: Thu, 26 Apr 2012 09:35:36 +0200
Subject: [PATCH 18/48] simplify an if condition in rt_prio

---
 include/linux/sched.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index 5e84053..ea87603 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1712,7 +1712,7 @@ static inline int dl_task(struct task_struct *p)
 
 static inline int rt_prio(int prio)
 {
-	if (unlikely(prio >= MAX_DL_PRIO && prio < MAX_RT_PRIO))
+	if ((unsigned)prio < MAX_RT_PRIO)
 		return 1;
 	return 0;
 }
-- 
1.7.9.5

