From 954417e606d51b629b3e7d806e2aef8c6e003a4b Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@gmail.com>
Date: Thu, 26 Apr 2012 09:36:47 +0200
Subject: [PATCH 19/48] Time's running.. update copyright year ;-)

---
 kernel/sched/dl.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/dl.c b/kernel/sched/dl.c
index fd75a65..dfd791a 100644
--- a/kernel/sched/dl.c
+++ b/kernel/sched/dl.c
@@ -9,7 +9,7 @@
  * than their reserved bandwidth will be slowed down (and may potentially
  * miss some of their deadlines), and won't affect any other task.
  *
- * Copyright (C) 2010 Dario Faggioli <raistlin@linux.it>,
+ * Copyright (C) 2012 Dario Faggioli <raistlin@linux.it>,
  *                    Juri Lelli <juri.lelli@gmail.com>,
  *                    Michael Trimarchi <michael@amarulasolutions.com>,
  *                    Fabio Checconi <fabio@gandalf.sssup.it>
-- 
1.7.9.5

