From 99d1e0f89ba65f5b7e6881c89299a75545d63cd5 Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@gmail.com>
Date: Wed, 21 Nov 2012 13:17:53 -0800
Subject: [PATCH 45/48] Fix compile error

---
 kernel/sched/deadline.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index 796af45..6b9879b 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -14,9 +14,10 @@
  *                    Michael Trimarchi <michael@amarulasolutions.com>,
  *                    Fabio Checconi <fchecconi@gmail.com>
  */
-#include <linux/math128.h>
 #include "sched.h"
 
+#include <linux/slab.h>
+
 struct dl_bandwidth def_dl_bandwidth;
 
 static inline struct task_struct *dl_task_of(struct sched_dl_entity *dl_se)
-- 
1.7.9.5

