From 3878c0fd2171fbb44f94a9e0e02d5d229cb7604b Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@gmail.com>
Date: Wed, 25 Apr 2012 22:52:37 +0200
Subject: [PATCH 17/48] Cut a redundant if statement in probe_wakeup

---
 kernel/trace/trace_sched_wakeup.c |    3 ---
 1 file changed, 3 deletions(-)

diff --git a/kernel/trace/trace_sched_wakeup.c b/kernel/trace/trace_sched_wakeup.c
index 680269b..8050c80 100644
--- a/kernel/trace/trace_sched_wakeup.c
+++ b/kernel/trace/trace_sched_wakeup.c
@@ -430,9 +430,6 @@ probe_wakeup(void *ignore, struct task_struct *p, int success)
 	tracing_record_cmdline(p);
 	tracing_record_cmdline(current);
 
-	if ((wakeup_rt && !rt_task(p)) ||
-			p->prio >= wakeup_prio ||
-			p->prio >= current->prio)
 	/*
 	 * Semantic is like this:
 	 *  - wakeup tracer handles all tasks in the system, independently
-- 
1.7.9.5

