From dac0db2eb5f0a4c7b25f68d171fbce5fc720cf28 Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@gmail.com>
Date: Wed, 2 May 2012 12:15:17 +0200
Subject: [PATCH 23/48] remove a redundant setscheduler call

---
 kernel/sched/core.c |    1 -
 1 file changed, 1 deletion(-)

diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 26c20c0..ee61203 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -4074,7 +4074,6 @@ recheck:
 
 	oldprio = p->prio;
 	prev_class = p->sched_class;
-	__setscheduler(rq, p, policy, param->sched_priority);
 	if (dl_policy(policy)) {
 		__setparam_dl(p, param);
 		__setscheduler(rq, p, policy, param->sched_priority);
-- 
1.7.9.5

