From 7a09dde873c0fe5085f9131eec85466bde3cff48 Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@gmail.com>
Date: Thu, 25 Oct 2012 12:08:00 -0700
Subject: [PATCH 38/48] sched: change Fabio Checconi's email address

---
 kernel/sched/deadline.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index afa6500..4608e39 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -12,7 +12,7 @@
  * Copyright (C) 2012 Dario Faggioli <raistlin@linux.it>,
  *                    Juri Lelli <juri.lelli@gmail.com>,
  *                    Michael Trimarchi <michael@amarulasolutions.com>,
- *                    Fabio Checconi <fabio@gandalf.sssup.it>
+ *                    Fabio Checconi <fchecconi@gmail.com>
  */
 #include <linux/math128.h>
 #include "sched.h"
-- 
1.7.9.5

