From 4ec76dc52224a14b6d4e3d956e389af1556430fa Mon Sep 17 00:00:00 2001
From: Juri Lelli <juri.lelli@gmail.com>
Date: Wed, 21 Nov 2012 13:10:25 -0800
Subject: [PATCH 44/48] sched: don't use u128 math

---
 kernel/sched/deadline.c |   11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index 805e61e..796af45 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -363,7 +363,7 @@ static void replenish_dl_entity(struct sched_dl_entity *dl_se,
 static bool dl_entity_overflow(struct sched_dl_entity *dl_se,
 			       struct sched_dl_entity *pi_se, u64 t)
 {
-	u128 left, right;
+	u64 left, right;
 
 	/*
 	 * left and right are the two sides of the equation above,
@@ -378,13 +378,10 @@ static bool dl_entity_overflow(struct sched_dl_entity *dl_se,
 	 * to the (absolute) deadline. Therefore, overflowing the u64
 	 * type is very unlikely to occur in both cases.
 	 */
-	left = mul_u64_u64(pi_se->dl_period, dl_se->runtime);
-	right = mul_u64_u64((dl_se->deadline - t), pi_se->dl_runtime);
+	left = pi_se->dl_period * dl_se->runtime;
+	right = (dl_se->deadline - t) * pi_se->dl_runtime;
 
-	if (cmp_u128(left, right) > 0)
-		return true;
-	
-	return false; 
+	return dl_time_before(right, left);
 }
 
 /*
-- 
1.7.9.5

